src/vmm/src/vstate/vcpu/aarch64.rs

```diff
@@ -105,36 +109,58 @@ impl KvmVcpu {
         // This reads back the kernel's preferred target type.
         vm_fd
             .get_preferred_target(&mut kvi)
-            .map_err(Error::VcpuPreferredTarget)?;
+            .map_err(Error::GetPreferredTarget)?;
         // We already checked that the capability is supported.
         kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_PSCI_0_2;
         // Non-boot cpus are powered off initially.
         if self.index > 0 {
             kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_POWER_OFF;
         }
 
-        self.fd.vcpu_init(&kvi).map_err(Error::VcpuInit)?;
+        self.fd.vcpu_init(&kvi).map_err(Error::Init)?;
         arch::aarch64::regs::setup_boot_regs(
             &self.fd,
             self.index,
             kernel_load_addr.raw_value(),
             guest_mem,
         )
-        .map_err(Error::REGSConfiguration)?;
+        .map_err(Error::ConfigureRegisters)?;
 
-        self.mpidr = arch::aarch64::regs::read_mpidr(&self.fd).map_err(Error::REGSConfiguration)?;
+        self.mpidr =
+            arch::aarch64::regs::read_mpidr(&self.fd).map_err(Error::ConfigureRegisters)?;
 
         Ok(())
     }
 
     /// Save the KVM internal state.
     pub fn save_state(&self) -> Result<VcpuState> {
-        Err(Error::UnsupportedAction("Saving the state"))
+        let mut state = VcpuState::default();
+
+        // Get this vCPUs multiprocessing state.
+        state.mp_state = arch::regs::get_mpstate(&self.fd).map_err(Error::SaveState)?;
+
+        arch::regs::save_core_registers(&self.fd, &mut state.core_regs)
+            .map_err(Error::SaveState)?;
+
+        arch::regs::save_system_registers(&self.fd, &mut state.sys_regs)
+            .map_err(Error::SaveState)?;
+
+        state.mpidr = arch::aarch64::regs::read_mpidr(&self.fd).map_err(Error::SaveState)?;
+
+        Ok(state)
     }
 
     /// Use provided state to populate KVM internal state.
-    pub fn restore_state(&self, _state: &VcpuState) -> Result<()> {
-        Err(Error::UnsupportedAction("Restoring the state"))
+    pub fn restore_state(&self, state: &VcpuState) -> Result<()> {
```

Comment:
Shouldn't we restore MPIDR as well?

Reply:
MPIDR is saved here because we need it in order to save the GIC's registers. This functionality is needed when saving/restoreing vm state (next PR).

Reply:
Actually, we restore the MPDIR at 151 line. Sorry for the wasted cycles. :D I'll resolve this.